### PR TITLE
feat: Meeting Schedule Quickstart

### LINKED ISSUE #374: Meeting Scheduling quickstart

### PR DESCRIPTION
**Pull request by zepfred**
## Description of the change

This PR adds a new quickstart for the Meeting Scheduling problem.

[Screencast from 2024-04-08 13-07-14.webm](https://github.com/TimefoldAI/timefold-quickstarts/assets/11961659/c84e91cc-5deb-4c3e-b1b1-c6bdb0705035)

### Development

- [x] The changes have been covered with tests, if necessary.
- [x] You have a green build, with the exception of the flaky tests.
- [x] UI and JS files are fully tested, the user interface works for all modules affected by your changes (e.g., solve and analyze buttons).
- [x] The network calls work for all modules affected by your changes (e.g., solving a problem).
- [x] The console messages are validated for all modules affected by your changes.

### Code Review

- [x] This pull request includes an explanatory title and description.
- [x] The GitHub issue is linked.
- [x] At least one other engineer has approved the changes.
- [ ] After PR is merged, inform the reporter.

### CODE REVIEW

File: use-cases/meeting-scheduling/pom.xml [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+    <groupId>org.acme</groupId>
+    <artifactId>timefold-solver-quarkus-meeting-scheduling-quickstart</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+    <properties>
+        <maven.compiler.release>17</maven.compiler.release>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+
+        <version.io.quarkus>3.9.1</version.io.quarkus>
```
Comment (triceo) [inline use-cases/meeting-scheduling/pom.xml]: We're back to 3.8.3 for now.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import ai.timefold.solver.core.impl.util.MutableInt;
+import ai.timefold.solver.core.impl.util.Pair;
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java]: Let's not use internal classes from the solver.
If you need a `Pair`, write the record yourself - it's trivial.
The `MutableInt` is unnecessary.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                                MeetingAssignment::getMeeting))
+                .join(MeetingAssignment.class,
+                        equal((leftRequiredAttendance, rightRequiredAttendance, leftAssignment) -> rightRequiredAttendance
+                                .getMeeting(),
+                                MeetingAssignment::getMeeting),
+                        overlapping((attendee1, attendee2, assignment) -> assignment.getStartingTimeGrain().getGrainIndex(),
+                                (attendee1, attendee2, assignment) -> assignment.getStartingTimeGrain().getGrainIndex() +
+                                        assignment.getMeeting().getDurationInGrains(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex() +
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java]: the grainIndex+durationInGrains is repeated so often that I think it would warrant a getter on assignment.

**Review by triceo — COMMENTED** [review]
First quick look.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java [hunk 4]
```diff
@@ -333,4 +331,51 @@ private <T, L> void applyRandomValue(int count, List<T> values, L secondParam, P
...
+        }
+
+        public int hashCode() {
+            return Objects.hash(this.value);
+        }
+
+        public String toString() {
+            return this.value.toString();
+        }
+    }
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java]: Considering that this is a quickstart, let's refactor the code so that mutable references aren't necessary. Every time you have to ask yourself "why is this here?" and the answer is not directly related to the idea you want to show the users, remove the thing.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java [hunk 5]
```diff
@@ -333,4 +331,51 @@ private <T, L> void applyRandomValue(int count, List<T> values, L secondParam, P
...
+            this.value = value;
+        }
+
+        public Key_ key() {
+            return this.key;
+        }
+
+        public Value_ value() {
+            return this.value;
+        }
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java]: It's a record; these are not necessary.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/Meeting.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        if (isRequired) {
+            if (requiredAttendances.stream().noneMatch(r -> r.getPerson().equals(person))) {
+                requiredAttendances.add(new RequiredAttendance(id, this, person));
+            }
+        } else {
+            if (preferredAttendances.stream().noneMatch(r -> r.getPerson().equals(person))) {
+                preferredAttendances.add(new PreferredAttendance(id, this, person));
+            }
+        }
+    }
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/Meeting.java]: I see two options here:
- Either don't check if the person is already present.
- Or if you want to check for that, do something if the person is already present. Fail, perhaps.

Also, instead of this if switch, I'd rather turn this into two methods - add required attendant, add preferred attendant. See how the `isRequired` method parameter effectively makes it two different methods? There is no shared code inside of the method.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+    @JsonIgnore
+    public int calculateOverlap(MeetingAssignment other) {
+        if (startingTimeGrain == null || other.getStartingTimeGrain() == null) {
+            return 0;
+        }
+        // start is inclusive, end is exclusive
+        int start = startingTimeGrain.getGrainIndex();
+        int otherStart = other.startingTimeGrain.getGrainIndex();
+        int otherEnd = otherStart + other.meeting.getDurationInGrains();
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java]: Didn't we add a method for this?

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        if (otherEnd < start) {
+            return 0;
+        }
+        int end = start + meeting.getDurationInGrains();
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java]: Dtto.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        if (end < otherStart) {
+            return 0;
+        }
+        return Math.min(end, otherEnd) - Math.max(start, otherStart);
+    }
+
+    @JsonIgnore
+    public Integer getLastTimeGrainIndex() {
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java:114]: Arguably if the counterpart to this is called `getStarting...`, this one should be `getEnding...`. Or they should be first/last. But it should be mutually consistent.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIdentityInfo;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.ObjectIdGenerators;
+
+@JsonIdentityInfo(scope = TimeGrain.class, generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
+public class TimeGrain implements Comparable<TimeGrain> {
+
+    private static final Comparator<TimeGrain> COMPARATOR = Comparator.comparing(TimeGrain::getDayOfYear)
+            .thenComparingInt(TimeGrain::getStartingMinuteOfDay);
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:15]: Why is `TimeGrain` comparable and can we do anything to stop it?

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        int hourOfDay = startingMinuteOfDay / 60;
+        int minuteOfHour = startingMinuteOfDay % 60;
+        return (hourOfDay < 10 ? "0" : "") + hourOfDay
+                + ":" + (minuteOfHour < 10 ? "0" : "") + minuteOfHour;
+    }
+
+    @JsonIgnore
+    public String getDateTimeString() {
+        return DAY_FORMATTER.format(getDate()) + " " + getTimeString();
+    }
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java]: Do we need this anywhere? What for?

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+                new Pair<>(0.33f, 12),
+                new Pair<>(0.33f, 16));
+        durationGrainsCount.forEach(p -> applyRandomValue((int) (p.key() * meetings.size()), meetings,
+                m -> m.getDurationInGrains() == 0, m -> m.setDurationInGrains(p.value())));
+        // Ensure there are no empty duration
+        meetings.stream()
+                .filter(m -> m.getDurationInGrains() == 0)
+                .forEach(m -> m.setDurationInGrains(8));
+        // Attendants
+        MutableInt attendantCount = new MutableInt();
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java]: If you don't use lambdas so much, you won't have to complicate your code with mutable references.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+            this.value = value;
+        }
+
+        public Key_ key() {
+            return this.key;
+        }
+
+        public Value_ value() {
+            return this.value;
+        }
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java]: Records don't need these.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+    // ************************************************************************
+    // Hard constraints
+    // ************************************************************************
+
+    public Constraint roomConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEachUniquePair(MeetingAssignment.class,
+                equal(MeetingAssignment::getRoom),
+                overlapping(MeetingAssignment::getGrainIndex,
+                        assignment -> assignment.getGrainIndex() +
+                                assignment.getMeeting().getDurationInGrains()))
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java]: Don't we have a method for this? If not, let's add it. It's used very often. Please check the rest of the constraint provider too.

**Review by triceo — COMMENTED** [review]
Nice!

Besides the comments in-line, I suggest we slightly reduce the number of people in the list. It seems a bit too long once again.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java (hunk 10 above)
Comment (zepfred) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:15]: We can remove it by using a `LocalDateTime`.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java (hunk 11 above)
Comment (zepfred) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java]: Good point. It is only used for `TimeGrain::toString()` but can be removed.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java (hunk 10 above)
Comment (triceo) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:15]: Hmm, if you say we need it for joiners in constraints, then I'm fine with `TimeGrain` being comparable. (Better do it directly instead of exposing local date and doing it there.) I was just making sure we're not carrying over some old cruft.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java (hunk 7 above)
Comment (zepfred) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java]: No, we did not.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java (hunk 14 above)
Comment (zepfred) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java]: We have a slightly different method for calculating the index. Now, we have a method for calculating the duration.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java [hunk 15]
```diff
@@ -99,23 +105,30 @@ public int calculateOverlap(MeetingAssignment other) {
...
+        return duration - 1;
+    }
+
+    @JsonIgnore
+    public int getEndingTimeGrainDuration() {
+        if (startingTimeGrain == null) {
+            return 0;
+        }
+        return startingTimeGrain.getGrainIndex() + meeting.getDurationInGrains();
     }
```
Comment (triceo) [inline use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java]: Now this is confusing.

`lastTimeGrainIndex` makes perfect sense - it is the last index. But `endingTimeGrainDuration` is also a grain index; arguably it's not a duration at all, because it starts from zero, not from the starting time grain.

The difference between the two is effectively `-1`. And that, arguably, isn't worth the confusion, or worth us sitting down and figuring out a better name for the method. Therefore my proposal:

- Keep `lastTimeGrainIndex` as that's consistent with `firstTimeGrainIndex`.
- Make `endingTimeGrainDuration` be `lastTimeGrainIndex + 1`.
- Inline all uses of `endingTimeGrainDuration`, removing the method entirely. Can be done with IntelliJ refactoring tools.

**Review by triceo — APPROVED** [review]
One more inline comment. (Sorry if I've changed my mind; I honestly didn't foresee the consequences of my original ask.)

Besides that:

- The red line is still there in the visualization.
- Any chance we could only display "reasonable" business hours in the view? Say 8:00 to 20:00? If not, nevermind.

After that, LGTM.

File: use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java (hunk 10 above)
Comment (zepfred) [reply use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:15]: Let me undo the change.

**Comment by zepfred** [comment]
> One more inline comment. (Sorry if I've changed my mind; I honestly didn't foresee the consequences of my original ask.)
> 
> Besides that:
> 
> * The red line is still there in the visualization.

Fixed.

> * Any chance we could only display "reasonable" business hours in the view? Say 8:00 to 20:00? If not, nevermind.

Done.
